docs/nebula-design-concepts.md

```diff
@@ -0,0 +1,216 @@
+and client console.
+
+![image][example5]
+
+### Storage layer
+
+The corresponding process of storage layer is nebula-storaged, and its core is the
+distributed Key-value Storage based on RAFT protocol. Currently it supports
+multiple storage engines like Rocksdb and HBase, etc.
+
+Raft protocol ensures data consistency  by leader/follower mechanism. Based on that, Nebula Storage makes the following optimizations:
+
+- Parallel Raft
+
+      Partitions of same ID from multiple machines form a raft group. And the concurrent operations are implemented with multiple sets of Raft groups.
+
+- Write Path & batch
+
+      In Raft protocol, multi-machine synchronization depends on log ID orders, which
+  lead to low throughput. Nebula achieves high throughput by batch and out-of-order
+  submission.
+
+- Copy learner based on synchronism
+
+      When added to cluster, new machines will be tagged as learner and pull data
+  from leader/follower synchronically. The learner will be tagged as follower and
+  participate in Raft protocol when it catches up leader.
+
+- Load-balance
+
+      For machines with high stress, migrate the partition they serve to machines
+  with  low stress for better load balancing.
+
+![image][example6]
+
+### Metadata service
+
+The process corresponding to the Meta service is nebula-metad, and its main functions are:
+
+-  User management
+
+      Nebula Graph's user system includes God user, Admin, User, and Guest, each is different in operating privileges.
+
+- Cluster configuration management
+
+      Support online and offline new servers.
+
+- Graph space management
+
+      Add, drop graph space, modify graph space configuration (Raft copies)
+
+-   Schema management
+
+      Nebula Graph is a strong schema database.
+
+      - Types of Tag and Edge properties are recorded by Meta service. Supported types are: int, double, timestamp, list, etc.
+
+      - Multi-version management, supporting adding, modifying and deleting schema, and recording its version.
+
+      - TTL (time-to-live) management, supporting automatic data deletion and space reclamation.
+
+The Meta Service layer is state, and just like the Storage layer, it maintains state persistence through KV Store.
+
+![image][example7]
+
+### Query Engine & Query Language (nGQL)
+
+The process corresponding to the computation layer is nebula-graphd, which consists
+of computing vertices that are fully equal, stateless, and unrelated. There is no
+communication among the computing vertices.
+
+The main function of the Query Engine layer is to analysis the nGQL text sent by the client and generate an execution plan through lexical analysis (Lexer) and parsing (Parser). Then the execution plan will pass to the the execution engine. The execution engine obtains the schema of the graph vertices and edges through the Meta Service, then obtains the data of the vertices and edges through the storage engine layer.
+
```

Comment:
It accepts the message from the client and generates the query plan after lexical analysis (Lexer), parsing (Parser) and query optimization. Then execution plan will be passed to the query execution engine. The query execution engine takes the query plans and interacts with metaserver and the storage engine to retrieve the schema and data.

Reply:
> e query execution engine takes the query

I guess you meant this part goes to the next session ;-)